pkg/dns/client/dns.go

```diff
@@ -388,14 +392,27 @@ func (h *LocalDNSServer) Close() {
 	}
 }
 
+func (h *LocalDNSServer) upstreamIterOffset(scope *istiolog.Scope) int {
+	if h.randomSelectUpstream {
+		offset, err := rand.Int(rand.Reader, big.NewInt(int64(len(h.resolvConfServers))))
+		if err != nil {
+			scope.Warnf("failed to select random iteration offset: %v", err)
+		}
+		return int(offset.Int64())
+	}
+	return 0
+}
+
 func (h *LocalDNSServer) queryUpstream(upstreamClient *dns.Client, req *dns.Msg, scope *istiolog.Scope) *dns.Msg {
 	if h.forwardToUpstreamParallel {
 		return h.queryUpstreamParallel(upstreamClient, req, scope)
 	}
 
 	var response *dns.Msg
 
-	for _, upstream := range h.resolvConfServers {
+	offset := h.upstreamIterOffset(scope)
+	for idx := range h.resolvConfServers {
+		upstream := h.resolvConfServers[(idx+offset)%len(h.resolvConfServers)]
```

Comment:
What happens if this picks the same server that was tried previously (in previous attempts)? Would there be a possibility of missing few upstream servers and return failure to the user?

Reply:
All servers in the resolv.conf file are still tried; this only changes is where in the list the server starts at. It may be logically simpler to simply clone the array and shuffle it.

As you mentioned, this could result in the application getting unlucky and choosing the same server, leading to an eventual resolution error in the client. So, this would only eventually help if the application itself retries establishing the connection after a resolution error. In contrast, currently, the application would suffer regardless of how many times it retries since the first entry in the resolv.conf is tried first each time.

Reply:
> All servers in the resolv.conf file are still tried; this only changes is where in the list the server starts at. It may be logically simpler to simply clone the array and shuffle it.

I am questioning this part. How will all servers be tried if the random index is duplicated in the loop?

Reply:
> random index is duplicated in the loop

Not sure I follow. The upstream to try in the loop is `(idx+offset)%len(h.resolvConfServers)`. So, for servers A B C D, if the random offset is 2, the loop should try the servers in the order C D A B.

Reply:
I see. Ignore me. Offset is calculated once. I thought it is calculated every time. So only starting position is random